### PR TITLE
Added `IOMessageWithMarkers`

### PR DESCRIPTION
**Pull request by iaronaraujo**
#### What problem is this solving?

Created `IOMessageWithMarkers` to be initially used by the `product-price`. 
**PENDING README UPDATE**
#### How should this be manually tested?

[Workspace](https://productprice--storecomponents.myvtex.com/classic-shoes/p?skuId=35)

#### Screenshots or example usage

image

#### Type of changes

| ✔️  | Type of Change                                                                            |
| --- | ----------------------------------------------------------------------------------------- |
| \_  | Bug fix                                |
| ✔️  | New feature                        |
| \_  | Breaking change  |
| \_  | Technical improvements  |
| \_  | None of the above                                                                         |

### CODE REVIEW

File: react/IOMessageWithMarkers.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable react/display-name */
```
Comment (Klynger) [inline react/IOMessageWithMarkers.tsx]: You don't need this I think

File: react/IOMessageWithMarkers.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      id={label}
+      values={{
+        ...markerComponents,
+        ...values,
+      }}
+    />
+  )
+}
+
+export default React.memo(IOMessageWithMarkers)
```
Comment (Klynger) [inline react/IOMessageWithMarkers.tsx]:
```suggestion
const IOMessageWithMarkers: IOMessageWithMarkersType = React.memo(function IOMessageWithMarkers(props) {
  const {
    label,
    markers = [],
    componentName,
    values = {},
  } = props
  const CSS_HANDLES = useMemo(() => markers.map(marker => {
    return `${componentName}_${marker}`
  }), [markers])
  const handles = useCssHandles(CSS_HANDLES)
  
  const markerComponents = markers.reduce((acc: any, marker) => {
    acc[marker] = (...chunks: any) => (
      <span className={handles[`${componentName}_${marker}`]}>{chunks}</span>
    )
    return acc
  }, {})

  return (
    <IOMessage
      id={label}
      values={{
        ...markerComponents,
        ...values,
      }}
    />
  )
})

export default IOMessageWithMarkers
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Added
+- `IOMessageWithMarkers`.
```
Comment (kaisermann) [inline CHANGELOG.md]: I'd add a small description of what `IOMessageWithMarkers` is here

File: react/IOMessageWithMarkers.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    return `${componentName}_${marker}`
+  }), [markers])
+  const handles = useCssHandles(CSS_HANDLES)
+  
+  const markerComponents = markers.reduce((acc: any, marker) => {
+    acc[marker] = (...chunks: any) => (
+      <span className={handles[`${componentName}_${marker}`]}>{chunks}</span>
+    )
+    return acc
+  }, {})
```
Comment (victorhmp) [inline react/IOMessageWithMarkers.tsx]: The type for `acc` and `markerComponents` could be `Record<string, ReactElement>`.
Also, I don't quite understand what is `...chunks`

File: react/__tests__/IOMessageWithMarkers.test.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    const label = '<bold>Hello World!</bold>'
+    const markers = ['bold']
+    const componentName = 'test'
+    const { getByText } = renderIOMessageWithIntl({ label, markers, componentName })
+    const element = getByText('Hello World!')
+
+    expect(element.className).toBe('test_bold')
+  })
+
+  it('check messaage with many markers and values', () => {
```
Comment (victorhmp) [inline react/__tests__/IOMessageWithMarkers.test.tsx]:
```suggestion
  it('check message with many markers and values', () => {
```

Comment (kaisermann) [inline react/__tests__/IOMessageWithMarkers.test.tsx]:
```suggestion
  it('check message with many markers and values', () => {
```

File: react/__tests__/IOMessageWithMarkers.test.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    const markers = ['bold', 'italic', 'pretty', 'sale']
+    const { getByText } = renderIOMessageWithIntl({ label, values, markers, componentName })
+
+    expect(getByText('A').className).toBe('test_bold')
+    expect(getByText('B').className).toBe('test_italic')
+    expect(getByText('C').className).toBe('test_pretty')
+    expect(getByText('D').className).toBe('test_sale')
+  })
+
+  // Right now this check fails for unkown reasons. This should be investigated more.
```
Comment (victorhmp) [inline react/__tests__/IOMessageWithMarkers.test.tsx]: Just is just to prevent merging

Comment (victorhmp) [reply react/__tests__/IOMessageWithMarkers.test.tsx]: Also, don't forget to remove `debug()` below :)

File: react/IOMessageWithMarkers.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable react/display-name */
+import React, { useMemo } from 'react'
+import { useCssHandles } from 'vtex.css-handles'
+
+import IOMessage from './IOMessage'
+import { IOMessageWithMarkers as IOMessageWithMarkersType } from './typings/IOMessageWithMarkers'
```
Comment (kaisermann) [inline react/IOMessageWithMarkers.tsx]: `.d.ts` files should not be directly imported. The project's `tsconfig.json` already includes all `.d.ts` inside the `typings` directory, you just need to use the type as if it's a global type.

File: react/IOMessageWithMarkers.tsx (hunk 2 above)
Comment (kaisermann) [reply react/IOMessageWithMarkers.tsx]: I think iaron's version is a tiny bit more readable, but both are ok

File: react/IOMessageWithMarkers.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+const IOMessageWithMarkers: IOMessageWithMarkersType = ({
+  label,
+  markers = [],
+  componentName,
+  values = {},
+}) => {
+  const CSS_HANDLES = useMemo(() => markers.map(marker => {
+    return `${componentName}_${marker}`
```
Comment (kaisermann) [inline react/IOMessageWithMarkers.tsx]: Did we discuss this handle naming pattern? Should we be using `_` as separators?

Comment (iaronaraujo) [reply react/IOMessageWithMarkers.tsx]: That's a great question

File: react/IOMessageWithMarkers.tsx (hunk 4 above)
Comment (klzns) [reply react/IOMessageWithMarkers.tsx]: chunks[ it's something from react-intl](https://github.com/formatjs/react-intl/blob/master/docs/Components.md#rich-text-formatting)

image

File: react/IOMessageWithMarkers.tsx (hunk 2 above)
Comment (klzns) [reply react/IOMessageWithMarkers.tsx]: I prefer how it is right now too

File: react/__tests__/IOMessageWithMarkers.test.tsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  it.skip('check markers with only one character', () => {
```
Comment (klzns) [inline react/__tests__/IOMessageWithMarkers.test.tsx]:
```suggestion
  it('check markers with only one character', () => {
```

File: react/__tests__/IOMessageWithMarkers.test.tsx [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+      a: 'A',
+      b: 'B',
+      c: 'C',
+      d: 'D'
+    }
+    const componentName = 'test'
+    const markers = ['a', 'bb', 'c', 'd']
+    const { getByText, debug } = renderIOMessageWithIntl({ label, values, markers, componentName })
+
+    debug()
```
Comment (klzns) [inline react/__tests__/IOMessageWithMarkers.test.tsx]:
```suggestion
```

File: docs/README.md [hunk 11]
```diff
@@ -1,3 +1,55 @@
...
+      class="Greetings_bold"
+    >
+      Hello,
+    </span>
+     John Doe!
+  </div>
+```
+
+
+**🔴 WARNING:** The `markers` and `values` props of this component are both passed to `IOMessage` through `values` so it is important to notive that you **should not** use the same labels within these props.
```
Comment (kaisermann) [inline docs/README.md]:
```suggestion
**🔴 WARNING:** The `markers` and `values` props of this component are both passed to `IOMessage` through `values` so it is important to notice that you **should not** use the same labels within these props.
```

File: docs/README.md [hunk 12]
```diff
@@ -1,3 +1,55 @@
...
+}
+```
+
+### IOMessage
+
+Coming soon :)
+
+### IOMessageWithMarkers
+
+This component is used to render messages with interpolations and custom tags (markers). These markers are useful to wrap a aprt of the message with CSS handles.
```
Comment (klzns) [inline docs/README.md]:
```suggestion
This component is used to render messages with interpolations and custom tags (markers). These markers are useful to wrap a part of the message with CSS handles.
```

File: docs/README.md [hunk 13]
```diff
@@ -1,3 +1,55 @@
...
+| `message` | `string` | The message that can contain values to be interpolated | `undefined` |
+| `markers` | [`string`] | The list of markers to be used during interpolation | `[]` |
+| `handleBase` | `string` | Handle that will be used along with the marker to create a span with `CSS handle` on the part wrapped by marks | `''` |
+| `values` | `Record<string, any>` | Values used in the message interpolation | `{}` |
+
+**Example:**
+ - message: '<bold>Hello,</bold> {name}!'
+ - markers: ['bold']
+ - handleBase: 'Greetings'
+ - values: { name: 'John Doe' }
```
Comment (klzns) [inline docs/README.md]: Change this to JS code

File: docs/README.md [hunk 14]
```diff
@@ -31,12 +31,12 @@ This component is used to render messages with interpolations and custom tags (m
 
 **Example:**
 ``` js
```
Comment (klzns) [inline docs/README.md]:
```suggestion
```jsx
```

**Pull request merged**